src/EditorFeatures/Core/Tagging/AbstractAsynchronousTaggerProvider.TagSource_ProduceTags.cs

```diff
@@ -259,18 +259,43 @@ private void EnqueueWork(bool highPriority, bool frozenPartialSemantics, Cancell
             bool calledFromJtfRun,
             CancellationToken cancellationToken)
         {
-            // Jump to the main thread, as we have to grab the spans to tag and the caret point.
-            await _dataSource.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken).NoThrowAwaitable();
+            // Note: this method is called in some blocking scenarios.  Specifically, when the outlining manager blocks
+            // on outlining tags.  As such, we use ConfigureAwait(true) and NoThrowAwaitable(captureContext: true) to
+            // ensure we're always coming back to the calling context as much as possible.  In the blocking case, this
+            // is good, so we don't have unnecessary thread switches.  In the non-blocking threadpool case, this is also
+            // fine as CA(true) will just keep us on the threadpool.
+
+            var isVisible = true;
+            var spansToTag = ImmutableArray<DocumentSnapshotSpan>.Empty;
+            SnapshotPoint? caretPosition = null;
+
+            // Enqueue work to a queue that will all tagger main thread work together in the near future. This let's
+            // us avoid hammering the dispatcher queue with lots of work that causes contention.  Additionally, use
+            // a no-throw awaitable so that in the common case where we cancel before, we don't throw an exception
+            // that can exacerbate cross process debugging scenarios.
+            await _dataSource._mainThreadManager.PerformWorkOnMainThreadAsync(() =>
+            {
+                _dataSource.ThreadingContext.ThrowIfNotOnUIThread();
+
+                isVisible = _visibilityTracker is null || _visibilityTracker.IsVisible(_subjectBuffer);
+                // Make a copy of all the data we need while we're on the foreground.  Then switch to a threadpool
+                // thread to do the computation. Finally, once new tags have been computed, then we update our state
+                // in a threadsafe fashion in the background.
+                spansToTag = GetSpansAndDocumentsToTag();
+                caretPosition = _dataSource.GetCaretPoint(_textView, _subjectBuffer);
+            }, cancellationToken).NoThrowAwaitable(captureContext: true);
+
+            // Since we don't ever throw above, check and see if the await completed due to cancellation and do not
+            // proceed.
             if (cancellationToken.IsCancellationRequested)
                 return null;
 
             // if we're tagging documents that are not visible, then introduce a long delay so that we avoid
-            // consuming machine resources on work the user isn't likely to see.  ConfigureAwait(true) so that if
-            // we're on the UI thread that we stay on it.
+            // consuming machine resources on work the user isn't likely to see.
```

Comment:
got rid of thsi comment as the concept around CA(true) is explained at the otp of the method.